upgradeability_with_vtable/contracts/Registry.sol

```diff
@@ -1,33 +1,57 @@
+ */
 contract Registry is IRegistry {
-  mapping (string => mapping (bytes4 => address)) implementations;
-
-  function addImplementationFromName(string version, string func, address impl) public {
-    return addImplementation(version, bytes4(keccak256(func)), impl);
+  // Mapping of versions to implementations of different functions
+  mapping (string => mapping (bytes4 => address)) internal versions;
+
+  /**
+  * @dev Registers a new version of a function with its implementation address
+  * @param version representing the version name of the new function implementation to be registered
+  * @param func representing the name of the function to be registered
+  * @param implementation representing the address of the new function implementation to be registered
+  */
+  function addVersionFromName(string version, string func, address implementation) public {
+    return addVersion(version, bytes4(keccak256(func)), implementation);
   }
 
-  function addImplementation(string version, bytes4 func, address impl) public {
-    require(implementations[version][func] == 0x0);
-    implementations[version][func] = impl;
-    ImplementationAdded(version, func, impl);
+  /**
+  * @dev Registers a new version of a function with its implementation address
+  * @param version representing the version name of the new function implementation to be registered
+  * @param func representing the signature of the function to be registered
+  * @param implementation representing the address of the new function implementation to be registered
+  */
+  function addVersion(string version, bytes4 func, address implementation) public {
+    require(versions[version][func] == address(0));
+    versions[version][func] = implementation;
+    VersionAdded(version, func, implementation);
   }
 
-  function getImplementation(string version, bytes4 func) public view returns (address) {
-    return implementations[version][func];
+  /**
+  * @dev Tells the address of the function implementation for a given version
+  * @param version representing the version of the function implementation to be queried
+  * @param func representing the signature of the function to be queried
+  * @return address of the function implementation registered for the given version
+  */
+  function getVersion(string version, bytes4 func) public view returns (address) {
+    return versions[version][func];
   }
 
-  function create(string version) public payable returns (Proxy) {
-    Proxy proxy = new Proxy(version);
-
+  /**
+  * @dev Creates an upgradeable proxy
+  * @return address of the new proxy created
+  */
+  function createProxy(string version, bytes4[] funcs) public payable returns (UpgradeabilityProxy) {
+    UpgradeabilityProxy proxy = new UpgradeabilityProxy(version, funcs);
     Upgradeable(proxy).initialize.value(msg.value)(msg.sender);
-
-    Created(proxy);
-
+    ProxyCreated(proxy);
     return proxy;
   }
```

Comment:
I don't like requiring the caller to keep track of all the `funcs` that should be included in a proxy for a version. Though it works for off-chain calls, invoking this method from another contract may be cumbersome.

WDYT about keeping an array of function names per version, and using it here in `createProxy`? 
State for the registry would then be:

```solidity
// Version -> function identifier -> implementation
mapping (string => mapping (bytes4 => address)) internal versions;

// Version -> function identifiers
mapping (string => bytes4[]) internal funcs;
```

Reply:
This comes after separating  the registration of new versions from upgradeability. It is important to analyze this scenario from that perspective.

That said, `funcs` is not to keep track of all the functions to be included in the proxy. Note that those are the functions to be upgraded before initializing, meaning they are the functions required for the initialization process itself.

Reply:
However, I think it is a good idea to have a mapping for the function identifiers